### ISSUE TITLE
Error Domain=NSCocoaErrorDomain Code=3840 "JSON text did not start with array or object and option to allow fragments not set."

### ISSUE THREAD

**Issue by smudireddy** [AzureAD/microsoft-authentication-library-for-objc#1156]
MSAL 1.1.11
ADFS + cert Based Auth + SSO extension distributed by MDM.

Able to login and use the App till token got expired. When token got expired (Token expiry configured Via conditional access policy "Grant" option), MSAL switches to broker extension  auth page and getting an error.

Error Domain=NSCocoaErrorDomain Code=3840 "JSON text did not start with array or object and option to allow fragments not set." UserInfo={NSDebugDescription=JSON text did not start with array or object and option to allow fragments not set., MSALCorrelationIDKey=29B66572-67E4-4E15-A6C0-DF34A695471A, MSALBrokerVersionKey=6.5.4}

Could you please check and help us what caused this issue. If user removes the account manually from Authenticator app and try manually remove the account then it started working again. With next token expiry it will again causing the issue.

**Comment by smudireddy** [comment]
With ADFS + Password case is working.

**Comment by smudireddy** [comment]
[Uploading MS_Authenticator_Logs.log…]()

**Comment by smudireddy** [comment]
https://docs.microsoft.com/en-us/azure/active-directory/conditional-access/howto-conditional-access-session-lifetime

**Comment by hieunguyenmsft** [comment]
@smudireddy can you reup the log file and the network trace? The link above doesn't work.
Thank you.

**Comment by smudireddy** [comment]
![File (1)](https://user-images.githubusercontent.com/13241928/99335281-07c60700-28a7-11eb-8f43-3c6009b299fe.jpg)

**Comment by smudireddy** [comment]
[MS_Authenticator_Logs.log](https://github.com/AzureAD/microsoft-authentication-library-for-objc/files/5550825/MS_Authenticator_Logs.log)

**Comment by hieunguyenmsft** [comment]
Thanks, I got the log. Let me check it and update you on anything I find.

**Comment by hieunguyenmsft** [comment]
@smudireddy There is an error when client trying to parse the response from server, I'm not sure it's from server side or client side. Can you help to provide a network trace? Thank you.

**Comment by smudireddy** [comment]
@hieunguyenmsft / @oldalton ,

[MaaS360_JSON  Error Proxy.chls.zip](https://github.com/AzureAD/microsoft-authentication-library-for-objc/files/5571922/MaaS360_JSON.Error.Proxy.chls.zip)

**Comment by hieunguyenmsft** [comment]
Thanks @smudireddy I'm working with our server team to investigate this issue. Will keep you posted.

**Comment by smudireddy** [comment]
@hieunguyenmsft  / @oldalton ,
Any update on this blocker issue ?

**Comment by rahulcode** [comment]
Any Updates here  ?

**Comment by smudireddy** [comment]
@oldalton ,
Any update on this issue from your side?.

**Comment by oldalton** [comment]
We know the root cause of the issue but it is hard to fix in the short term. Let's discuss priority of this offline.

**Comment by smudireddy** [comment]
Issue should be fixed otherwise users with Certificate based authentication case will not work when token got expired.

**Comment by aherciya** [comment]
A fix is in place for our next release(1.1.15).

**Comment by rahulcode** [comment]
Is this handled with the latest 1.1.15 release?

**Comment by oldalton** [comment]
Unfortunately we found some additional edge cases, so latest release doesn't yet fully address this. We'll keep looking into other possible workarounds for this case.

**Comment by rahulcode** [comment]
Thanks, @oldalton for the update.
Is there any update on this issue ?
Thanks

**Comment by oldalton** [comment]
@rahulcode  - could you try with the build number 6.5.51 and see if the issue is resolved? It should be currently in the App Store.

**Comment by rahulcode** [comment]
Thanks @oldalton for the update. 

1 Query before starting validation : 

Which Version of the MSAL library should be taken along with the latest Microsoft Authenticator Application (6.5.51)?
Will the fix be backward compatible i.e. Will it work with prior MSAL release versions, Or do we need to take in the latest version of MSAL + 6.5.51 Authenticator?

Thanks

**Comment by antrix1989** [comment]
Thanks @rahulcode , we will take a look.

**Comment by oldalton** [comment]
Sent a follow up email to get a network trace. Unable to reproduce with in-house CBA setup.

**Comment by rahulcode** [comment]
Hi @oldalton 
Hope you are doing good.

Tested again the complete workflows and following is the summary with all logs :

1) **Enroll Now**
**Flow to reproduce before :**
After Sign In Frequency is applied from CA, Try to fetch the token interactively.

Now, ( With Latest Authenticator) ( This is a new issue in Same workflow )
We are **not seeing JSON issue in the same workflow** , but now we are shown with the **Enroll now Screen**
After getting into this State fetching of new token is not possible until account is cleared from Authenticator Application.
So, this is now the blocking 

For which the logs of authenticator and Charles are here. 

[Logs_EnrollNowIssue_Json_Issue.zip](https://github.com/AzureAD/microsoft-authentication-library-for-objc/files/6349323/Logs_EnrollNowIssue_Json_Issue.zip)

  2) JSON Issue is seen only when calling remove account and is non impacting to any of our app workflows now.
During remove account MSAL library shows JSON issue but there are no network traces for the same as no call is made over the network.MSAL Logs for the same are attached in above zip.

Let me know if we need more information. We can schedule a call if more info needed.

**Comment by oldalton** [comment]
Thanks! I'll take a look at the "enroll now" issue and report back. 

However, just to confirm the status of the JSON issue - I think before this could be reproduced if you'd already have an authenticator account, and would call MSAL interactively with prompt=login, or if user's password would change and you were calling MSAL immediately after that. 

I'd like to confirm that this issue is fully resolved for you? Because I can still see it in the logs that were uploaded 3 weeks ago, but I'm not able to reproduce it locally. Trying to make sure that all edge cases are covered.

**Comment by rahulcode** [comment]
JSON issues we are not seeing now.
Tested the similar flows again.
Will test password change cases and update again.

Is there information/update on the Cert Based Issue after Sign In Frequency( No **Continue** button (as Enroll now screen is removed ) appears and user can authenticate properly ?)

**Comment by oldalton** [comment]
`Is there information/update on the Cert Based Issue after Sign In Frequency( No Continue button (as Enroll now screen is removed ) appears and user can authenticate properly ?)`

We have a fix in progress for it. I'll update this thread when it's available for public testing. Thanks for your patience!

**Comment by hieunguyenmsft** [comment]
User reported this is fixed in MSAL 1.1.19
